### PR TITLE
Define boolean constants in Clarity

### LINKED ISSUE #1342: Forbid binding boolean constants in Clarity
Clarity should forbid binding `true` and `false` as can be done with `let`:

```clojure
(let ((true 0)) true)   ;; evaluates to 0
(let ((false 1)) false) ;; evaluates to 1
```

See also: #1327

### PR DESCRIPTION
**Pull request by njordhov**
Defines global constants in Clarity for `true` and `false`.

This resolves #1342 and #1327.

### CODE REVIEW

File: src/vm/docs/mod.rs [hunk 1]
```diff
@@ -91,6 +91,27 @@ const NONE_KEYWORD: KeywordAPI = KeywordAPI {
 "
 };
 
+const TRUE_KEYWORD: KeywordAPI = KeywordAPI {
+    name: "true",
+    output_type: "(optional ?)",
```
Comment (kantai) [inline src/vm/docs/mod.rs]: Output type should be `bool`

File: src/vm/tests/costs.rs [hunk 2]
```diff
@@ -49,11 +49,11 @@ pub fn get_simple_test(function: &NativeFunctions) -> &'static str {
         AsMaxLen => "(as-max-len? list-bar u3)",
         Len => "(len list-bar)",
         ListCons => "(list 1 2 3 4)",
-        FetchEntry => "(map-get? map-foo {a 1})",
-        FetchContractEntry => "(contract-map-get? .contract-other map-foo {a 1})",
-        SetEntry => "(map-set map-foo {a 1} {b 2})",
-        InsertEntry => "(map-insert map-foo {a 2} {b 2})",
-        DeleteEntry => "(map-delete map-foo {a 1})",
+        FetchEntry => "(map-get? map-foo (tuple (a 1)))",
```
Comment (kantai) [inline src/vm/tests/costs.rs]: Why revert this to `(tuple ...)` syntax?

Comment (njordhov) [reply src/vm/tests/costs.rs]: Tuple literals here failed unit testing. Figuring out why was not within the scope of the PR but likely resolved with #1337 eliminating implicit tuples.

File: src/vm/docs/mod.rs (hunk 1 above)
Comment (njordhov) [reply src/vm/docs/mod.rs]: Good catch. Thanks. Resolved in 46f70fc.

Perhaps inconsistencies in documented output_type could be caught in unit testing.

File: src/vm/tests/costs.rs (hunk 2 above)
Comment (kantai) [reply src/vm/tests/costs.rs]: Gotcha -- this should be resolved via #1340, which just merged to `master`

Comment (njordhov) [reply src/vm/tests/costs.rs]: Seems to do the trick. File reverted to original tuple literals.

**Comment by kantai** [comment]
This looks great to me @njordhov -- thanks for the PR!

I think we should only support one syntax for true/false values -- using `true` and `false` is probably preferable to `'true` and `'false`, in which case, we should probably remove the support for that syntax from the parser.

I'd like to hear from some others on that as well (@lgalabru, @jcnelson)

File: src/vm/tests/simple_apply_eval.rs [hunk 3]
```diff
@@ -588,7 +588,9 @@ fn test_bool_functions() {
          "(and 'false (> 1 (/ 10 0)))",
          "(or 'true (> 1 (/ 10 0)))",
          "(or 'false 'false 'false)",
-         "(not 'true)"];
+         "(not 'true)",
```
Comment (jcnelson) [inline src/vm/tests/simple_apply_eval.rs]: Shouldn't this be `(not true)`?

**Comment by jcnelson** [comment]
Thanks for the PR @njordhov!

> I think we should only support one syntax for true/false values -- using true and false is probably preferable to 'true and 'false, in which case, we should probably remove the support for that syntax from the parser.

I was going to ask this very question after looking through the code.  As a general PL design principle, I agree with @kantai that we should have only one syntax for a thing instead of two.

**Comment by njordhov** [comment]
I agree that using the `true` and `false` constants are preferable to the quoted literals. I can take care of removing the quoted boolean syntax in the code. Should that be a separate issue or an extension to this one?

**Comment by kantai** [comment]
> Should that be a separate issue or an extension to this one?

An extension to this PR would be preferable, thanks

**Comment by njordhov** [comment]
There are two failing tests, but they seem to depend on external stored contracts that still use the legacy boolean literals:
```
failures:
    testnet::tests::should_succeed_handling_malformed_and_valid_txs
    testnet::tests::should_succeed_mining_valid_txs
```

**Comment by kantai** [comment]
> There are two failing tests, but they seem to depend on external stored contracts that still use the legacy boolean literals:
> failures:
>    testnet::tests::should_succeed_handling_malformed_and_valid_txs
>    testnet::tests::should_succeed_mining_valid_txs

Yep -- those tests include raw serialized contract-publish transactions with the legacy literals. @lgalabru would know best how to update those tests, but it may be better to replace those raw serializations with generated serializations (such as is in `vm::tests::integrations`)

**Comment by lgalabru** [comment]
Ah yes, this is some kind of end to end test. If you need to update these, you can either follow the instructions here:
https://github.com/blockstack/stacks-blockchain/blob/a1392072dbb87594e3ed9822955e33cbbc22f8d9/src/testnet/tests.rs#L41-L54
and replace the transactions bytes for each round, or follow the recommendation made by @kantai.

**Comment by kantai** [comment]
@njordhov -- the branch `njordhov-feature/boolean-constants` on this repo patches the end to end tests:

https://github.com/blockstack/stacks-blockchain/tree/njordhov-feature/boolean-constants

**Comment by njordhov** [comment]
> the branch njordhov-feature/boolean-constants on this repo patches the end to end tests:

Thanks. Merged. For the conflict in `blocks.rs`, which branch should be used?

**Comment by kantai** [comment]
> Thanks. Merged. For the conflict in blocks.rs reported by ci, which branch should be used?

If you pull https://github.com/blockstack/stacks-blockchain/tree/njordhov-feature/boolean-constants again, then it should resolve the conflicts.

Note that this is also why in the future, it's better not to make too many whitespace edits -- most of the conflicts were caused by that.

**Comment by njordhov** [comment]
There is only one remaining failure during ci unit testing, but the contract tested seems just fine:

```
failures:
    testnet::helium::tests::should_succeed_mining_valid_txs
```

**Comment by kantai** [comment]
> There is only one remaining failure during ci unit testing, but the contract tested seems just fine:
>
> failures:
>   testnet::helium::tests::should_succeed_mining_valid_txs


Yep -- it looks like another PR added some `print` statements to that contract that it was testing. I updated the branch `njordhov-feature/boolean-constants` with conflicts with master resolved, and that test case fixed so that it uses lazy_static to calculate the publish transaction bytes, which should prevent this kind of things going forward.

**Comment by njordhov** [comment]
@kantai Thanks. Merged and pushed. RTG?

File: src/vm/tests/simple_apply_eval.rs [hunk 4]
```diff
@@ -609,13 +613,17 @@ fn test_bad_lets() {
...
+        "(let ((true 0)) true)",
+        "(let ((false 1)) false)"];
 
     let expectations: &[Error] = &[
         CheckErrors::NameAlreadyUsed("tx-sender".to_string()).into(),
         CheckErrors::NameAlreadyUsed("*".to_string()).into(),
         CheckErrors::NameAlreadyUsed("a".to_string()).into(),
-        CheckErrors::NoSuchDataVariable("cursor".to_string()).into()];
+        CheckErrors::NoSuchDataVariable("cursor".to_string()).into(),
+        CheckErrors::NameAlreadyUsed("true".to_string()).into(),
```
Comment (lgalabru) [inline src/vm/tests/simple_apply_eval.rs:625]: Throwing `NameAlreadyUsed` is probably not optimal in terms of DX (but it's consistent with the errors currently in place). Should we create an issue for this @kantai?

Comment (kantai) [reply src/vm/tests/simple_apply_eval.rs:625]: Yep, let's open an issue

**Comment by kantai** [comment]
Great! Will merge

**Pull request merged**